remove old synclogs during form submission

snopoke opened the pull request:
##### SUMMARY
This restores functionality remove in https://github.com/dimagi/commcare-hq/pull/20119. The original functionality was removed because in SQL we were able to easily delete old synclogs by dropping old table partitions: https://github.com/dimagi/commcare-hq/blob/579b52f1037d30c8e9184214f7278bb1d6a8b9d2/corehq/ex-submodules/casexml/apps/phone/tasks.py#L93

On ICDS we are now the position that only dropping very old logs does not free up enough storage space so we want to re-introduce the purging of old logs after new synclogs become active.

This is discussed further in: https://dimagi-dev.atlassian.net/browse/IIO-447

@emord I decided to take this approach instead of batching it via a pillow since it's much simpler and because I don't think it introduces significantly more load since it's only adding a single delete on the first form after a sync.

Also this (https://github.com/dimagi/commcare-hq/pull/25596) will reduce the load on the synclog DB for ~every form submission.

If it turns out that this puts too much load on the DB we can move to the batched approach.


This PR also includes some cleanup. The main change is in e99f31e .

snopoke committed:
remove unnecessary check

snopoke committed:
remove legacy exception processing

This soft assert hasn't been fired for a very long time

snopoke committed:
cleanup related code

snopoke committed:
simplify delete method

snopoke committed:
delete previous synclog during form submission

snopoke committed:
update sql fields

snopoke committed:
remove unnecessary function + params

snopoke committed:
add test

emord commented on corehq/ex-submodules/casexml/apps/case/util.py:
```diff
@@ -110,43 +110,12 @@ def get_case_xform_ids(case_id):
-            update_sync_log_with_checks(updated_log, xform, cases, case_db,
-                                        case_id_blacklist=case_id_blacklist)
+def prune_previous_log(sync_log):
+    if sync_log.previous_log_id:
```
> If it turns out that this puts too much load on the DB we can move to the batched approach.

Since this is a concern, should we gate this with a localsetting that defaults to true (or a feature flag I suppose)? that way we can quickly firefight without a code change.

snopoke replied:
bcfe49dc49e

snopoke commented:
I thought of another way to do the batch deletion in SQL. Not sure it will be very efficient though:
```sql
DELETE FROM phone_synclogsql WHERE synclog_id IN (
  WITH RECURSIVE synclog_cte (synclog_id, previous_synclog_id, user_id, last_submitted, date, level)
  AS (
      -- find most recent synclog that we know been submitted against
      SELECT synclog_id, previous_synclog_id, user_id, last_submitted, date, level FROM (
        SELECT
            ROW_NUMBER() over w as row,
            synclog_id,
            previous_synclog_id,
            user_id,
            last_submitted,
            date,
            0 as level
        FROM phone_synclogsql
        WHERE last_submitted IS NOT NULL
        WINDOW w AS (
          PARTITION BY user_id
           ORDER BY date desc RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING
      ) as a WHERE row = 1
      UNION ALL
      -- find all child synclogs
      SELECT
          child.synclog_id,
          child.previous_synclog_id,
          child.user_id,
          last_submitted,
          child.date,
          1 + parent.level as level
      FROM synclog_cte parent
      JOIN phone_synclogsql child ON child.synclog_id = parent.previous_synclog_id
  )
  SELECT synclog_id FROM synclog_cte WHERE level > 0  --exclude the first one
);
```

snopoke committed:
Merge branch 'master' into sk/synclog-purge

snopoke committed:
add toggle to allow controlled rollout

emord reviewed (approved):
looks like there's an error with how the featureflag is called, but approving the approach

snopoke committed:
pass in toggle namespace

snopoke committed:
set last_submitted if it was not already set

This makes sure that every synclog which has received
a submission get's this field set. This will allow
more efficient deletion of synclogs.

snopoke committed:
Merge branch 'master' into sk/synclog-purge

snopoke committed:
enable flag in test
